Feature/1660 SqlUtil op_in()

gamato opened the pull request:
Implements a work around for issue #1660

gamato committed:
SqlUtil: add tests for op_in() of empty and too big lists

The latter checks limitations of Oracle (ORA-22165 or ORA-01795).

gamato committed:
OracleSqlUtil: work around ORA-22165 from op_in() with a too big list

The trick here is to split long arrays to chunks of no more than 32767 elements.

gamato committed:
SqlUtil: update release notes for issue #1660

davidnich commented on qlib/OracleSqlUtil.qm:
```diff
@@ -1849,21 +1852,32 @@ my any $v = $c.name;
 
-                        return cn + " in (select column_value from table(%v))";
+                        if (count)
+                            return cn + " in (" + (map "select column_value from table(%v)", range(1,count)).join(" union all ") + ")";
```
please use xrange() rather than range() here: https://github.com/qorelanguage/qore/wiki/ProTips#xrange-vs-range

davidnich replied:
please use `xrange()` rather than `range()` here: https://github.com/qorelanguage/qore/wiki/ProTips#xrange-vs-range

davidnich commented on qlib/OracleSqlUtil.qm:
```diff
@@ -255,6 +255,9 @@ my hash $schema = (
 
     @section ora_relnotes Release Notes
 
+    @subsection v122 OracleSqlUtil v1.2.2
```
please also update the module's version string in the module header to "1.2.2"

davidnich commented on doxygen/lang/900_release_notes.dox.tmpl:
```diff
@@ -133,6 +133,7 @@
       - added logic to attempt to mask passwords in debug log messages (<a href="https://github.com/qorelanguage/qore/issues/1086">issue 1086</a>)
     - <a href="../../modules/SqlUtil/html/index.html">SqlUtil</a> module fixes:
       - fixed a bug in update and upsert statement generation when the given data does not have enough columns to use the unique index found, an error message is generated that contains all the columns names instead of just the column names required by the index (<a href="https://github.com/qorelanguage/qore/issues/1013">issue 1013</a>)
+      - worked around ORA-22165 from op_in() caused by Oracle's limit on number of collection elements (<a href="https://github.com/qorelanguage/qore/issues/1660">issue 1660</a>)
```
please put this line under the `OracleSqlUtil` module changes

gamato committed:
OracleSqlUtil: update #1660 to address feedback from David

omusil24 commented:
@gamato FreetdsSqlUtil.qtest fails because of this:
https://support.microsoft.com/en-us/help/913050/error-message-when-you-run-a-query-in-sql-server-2005-internal-error-an-expression-services-limit-has-been-reached

pvanek commented:
How it will work with orderby and / or aggregate functions? Also I suggest to extend docs to say that using huge in lists is a sign of potentially bad design of an application

davidnich commented:
@gamato as discussed, we need to first target this for 0.8.12

davidnich reviewed (commented):
as discussed, we need to first target this for 0.8.12

davidnich commented:
@gamato @pvanek I see two issues here, of which only one is clear to me how to solve:
1) SqlUtil should make `op_in()` work with arbitrarily large argument ranges if technically possible - no problem to document that it's bad practice to use too large lists here (but we should define how large the lists should be to make them reasonable and also why)
2) how to describe to users how to make queries selecting a large number of keys/ids known at runtime and not in the database - this is not clear to me how to do it if `op_in()` is not appropriate

gamato committed:
SqlUtil: skip too long list test for op_in() for FreetdsSqlUtil

As we do not have any workaround this test case would fail depending on MS SQL version.

davidnich commented:
@gamato please retarget this pull request to 0.8.12 as discussed

gamato committed:
SqlUtil: add notes about possible op_in() arg list limitations
